src/models/thread.ts

```diff
@@ -377,15 +378,15 @@ export class Thread extends ReadReceipt<ThreadEmittedEvents, ThreadEventHandlerM
             // When there's no server-side support, just add it to the end of the timeline.
             this.addEventToTimeline(event, toStartOfTimeline);
             this.client.decryptEventIfNeeded(event);
-        } else if (!toStartOfTimeline && this.initialEventsFetched && isNewestReply) {
+        } else if (event.isRelation(RelationType.Annotation) || event.isRelation(RelationType.Replace)) {
+            this.addRelatedThreadEvent(event, toStartOfTimeline);
+            return;
+        } else if (!toStartOfTimeline && isNewestReply) {
             // When we've asked for the event to be added to the end, and we're
             // not in the initial state, and this event belongs at the end, add it.
             this.addEventToTimeline(event, false);
             this.fetchEditsWhereNeeded(event);
-        } else if (event.isRelation(RelationType.Annotation) || event.isRelation(RelationType.Replace)) {
-            this.addRelatedThreadEvent(event, toStartOfTimeline);
-            return;
-        } else if (this.initialEventsFetched) {
+        } else {
             // If initial events have not been fetched, we are OK to throw away
```

Comment:
we're no longer checking if initial events have been fetched before hitting this code?

Reply:
Correct: this is the main part of the fix. We previously checked whether initial events had been fetched and, if they hadn't threw the event we were going to add in the bin, in favour of waiting for it to arrive from the pagination. This is what caused the races, because any code that then ran to check the state of the thread would find that it was empty.

Reply:
ok, but doesn't that mean that this comment is no longer relevant?

Also, I haven't fully grokked the change or the comment, but it feels like the block comment between lines 92 and 123 might need a re-visit?

Reply:
> Also, I haven't fully grokked the change or the comment, but it feels like the block comment between lines 92 and 123 might need a re-visit?

On a more careful reading, I think the comment on `lastEvent` is still valid.

I still think that *this* comment is outdated.

Reply:
Oh, haha, yes, I see what you mean. Fixed the comment (e2949fcde), and the block comment (059caff35).